packages/go_router/lib/src/route.dart

```diff
@@ -5,32 +5,141 @@
+  ///             final Family family = Families.family(state.params['fid']!);
+  ///             return MaterialPage<void>(
+  ///               key: state.pageKey,
+  ///               child: FamilyPage(family: family),
+  ///             );
+  ///           },
+  ///           routes: <GoRoute>[
+  ///             GoRoute(
+  ///               path: 'person/:pid',
+  ///               pageBuilder: (BuildContext context, GoRouterState state) {
+  ///                 final Family family = Families.family(state.params['fid']!);
+  ///                 final Person person = family.person(state.params['pid']!);
+  ///                 return MaterialPage<void>(
+  ///                   key: state.pageKey,
+  ///                   child: PersonPage(family: family, person: person),
+  ///                 );
+  ///               },
+  ///             ),
+  ///           ],
+  ///         ),
+  ///       ],
+  ///     ),
+  ///   ],
+  /// );
+  ///
+  /// If there are multiple routes that match the location, the first match is used.
+  /// To make predefined routes to take precedence over dynamic routes eg. '/:id'
+  /// consider adding the dynamic route at the end of the routes
+  /// For example:
+  /// ```
+  /// final GoRouter _router = GoRouter(
+  ///   routes: <GoRoute>[
+  ///     GoRoute(
+  ///       path: '/',
+  ///       redirect: (_) => '/family/${Families.data[0].id}',
+  ///     ),
+  ///     GoRoute(
+  ///       path: '/family',
+  ///       pageBuilder: (BuildContext context, GoRouterState state) => ...,
+  ///     ),
+  ///     GoRoute(
+  ///       path: '/:username',
+  ///       pageBuilder: (BuildContext context, GoRouterState state) => ...,
+  ///     ),
+  ///   ],
+  /// );
+  /// ```
+  /// In the above example, if /family route is matched, it will be used.
+  /// else /:username route will be used.
+  ///
+  /// See [Sub-routes](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/sub_routes.dart)
+  /// for a complete runnable example.
+  final List<RouteBase> routes;
+}
+
+/// A route that is displayed visually above the matching parent route using the
+/// [Navigator].
+///
+/// The widget returned by [builder] is wrapped in [Page] and provided to the
+/// root Navigator, the nearest ShellRoute ancestor's Navigator, or the
+/// Navigator with a matching [parentNavigatorKey].
+///
+/// The Page depends on the application type: [MaterialPage] for
+/// [MaterialApp], [CupertinoPage] for [CupertinoApp], or
+/// [NoTransitionPage] for [WidgetsApp].
+class GoRoute extends RouteBase {
+  /// Constructs a [GoRoute].
+  /// - [path] and [name] cannot be empty strings.
+  /// - One of either [builder] or [pageBuilder] must be provided.
   GoRoute({
     required this.path,
     this.name,
+    this.builder,
     this.pageBuilder,
-    this.builder = _invalidBuilder,
-    this.routes = const <GoRoute>[],
-    this.redirect = _noRedirection,
+    this.parentNavigatorKey,
```

Comment:
We should probably has some guard around using parentNavigatorKey that is not in the ancester navigator chain.
For example we don't want this.
```
GoRoute(
   routes: [
      ShellRoute(navKey: a, routes:...),
      ShellRoute(
          navKey:b,
          routes: [GoRoute(parentNavKey: a)]
      )
   ]
)
```

Or something like this
```
GoRoute(
   routes: [
      ShellRoute(
          navKey:a,
          routes: [
             GoRoute(
                 parentNavKey: root,
                 routes: [GoRoute(parentNavKey: a)]
             )
          ]
      )
   ]
)
```

Reply:
85a1a6d6cb5a7512999903e05dd0d98694169c1e - I'm not sure what the second example is supposed to test. Is the rule that parentNavigatorKey cannot be specified if the parent GoRoute has a parentNavigatorKey?

Reply:
73f47474c8cf057d909941b7c22ac6b52cd509cb